keyvi/bin/keyvi_c/c_api.cpp

```diff
@@ -0,0 +1,185 @@
+/* * keyvi - A key value store.
+ *
+ * Copyright 2015, 2016 Narek Gharibyan <narekgharibyan@gmail.com>
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+/*
+ * c_api.cpp
+ *
+ *  Created on: September 4, 2017
+ *      Author: Narek Gharibyan <narekgharibyan@gmail.com>
+ */
+
+
+#include <iostream>
+
+#include "c_api/c_api.h"
+#include "dictionary/dictionary.h"
+#include "dictionary/completion/prefix_completion.h"
+#include "dictionary/completion/multiword_completion.h"
+
+
+using namespace keyvi::dictionary;
+
+namespace {
+    char *
+    std_2_c_string(const std::string &str) {
+        auto result = static_cast<char *>(malloc(str.size() + 1));
+        strcpy(result, str.c_str());
```

Comment:
although trivial: I suggest to use strncpy

Reply:
Can you please elaborate a bit on this ? I thought it's safe as I'm using `std::string::c_str()`

https://stackoverflow.com/questions/1258550/why-should-you-use-strncpy-instead-of-strcpy. Please check all answers.

Reply:
security best practice: if you can do not use any unbound functions like strcpy. 

As said trivial: it is obviously safe in this case. But as said, it is better to simply avoid it if you can. 

<anecdote>
Bigger companies (e.g. microsoft) disallow to write such code enforced by static analysis, you have to write a long special formatted docstring with an exception why you to do it like this to get this in.
</anecdote>

Long story short, it's easy to avoid here, so lets do so.

Theoretically it is also faster, strcpy involves calculating the size of the string, strncpy does not need to.

Reply:
of course `\0` termination must be assured, and just saying not to use strcpy if possible, therefore any alternative is fine.